### ISSUE TITLE
Froala Editor Replaces and Strips out HTML

### ISSUE THREAD

**Issue by nickian** [mautic/mautic#1959]
The new Froala editor in Mautic 2.0 seems very problematic. You can't simply paste in custom HTML without the editor radically modifying your code. I typically use Foundation for Emails (http://foundation.zurb.com/emails.html) and the Froala Editor completely destroys these templates because it "beautifies" the code and doesn't accept certain tags.

Is there a way to set this editor to simply accept the raw input I give it? I don't want it to strip or change anything I paste into the code view. After a glance at their documentation, I don't see an option for this.

The new email templates are nice, but this is going to be a deal breaker for people who are serious about highly customized HTML email designs that are compatible with as many email clients as possible.

**Comment by escopecz** [comment]
Could you please open this issue at https://github.com/froala/wysiwyg-editor/issues ? We can specifically list the tags and attributes which are allowed, but we always miss some. Froala also strips the comments which are used often in email themes. But Mautic dev team can hardly do something about it.

**Comment by nickian** [comment]
I will open an issue on their repository. If this can't be resolved by configuring Froala, I would suggest that Mautic should only use Froala as a WYSIWYG editor or in the builder. There should be a separate tab with a just straight CodeMirror or Ace Editor for saving raw, custom HTML.

**Comment by escopecz** [comment]
Funny you say that. It was just like this till the last moment before releasing 2.0. Many people complained that they don't want to use the builder and they don't know HTML. So we replaced the CodeMirror editor with Froala with the CodeMirror switch option.

Froala and CodeMirror cannot co-exist on the edit form, because we have to make sure that if you edit something in Froala, the change will appear in all the other areas where you can edit the content. So Froala would still load and destroy your HTML even if you would not edit the content in it.

**Comment by nickian** [comment]
I totally understand why some people would want a WYSIWYG, but you can't have your cake and eat it too. In other words, if you truly want your email to look good across all email clients and devices, you _have_ to know HTML, and more specifically, you have to know HTML _for email_, which requires using very specific code and certain hacks to compensate for different HTML rendering engines (e.g., Outlook).

WYSISWYG editing is great the Mom and Pop shop, and there's nothing wrong with having the option, but professional designers and agencies need to be able to use their own custom code. For example, Pardot, Hubspot, Campaign Monitor, MailChimp, Sendy, etc. – all of these platforms allow you to import your own custom code, unobstructed, for good reason.

I opened an issue on Froala's Github (https://github.com/froala/wysiwyg-editor/issues/1386) to see if there's a simple way to configure the editor to accept raw code.

**Comment by escopecz** [comment]
Hey, I'm totally with you on this one. But I also understand that not all marketers wants to learn HTML. So the only solution I can think of is to let each Mautic user choose which editor they want to use in their profile. What do you think? Can you think of a better solution?

The downside is that if your colleague opens and saves the email you created in the CodeMirror in the WYSIWYG editor, he "clears" your code again. So maybe we'll have to lock the Content tab for some users which means a new permission for this purpose.

**Comment by nickian** [comment]
Hopefully there is just a way to prevent Froala from stripping code. Once the template has been pasted into the code view, it would be nice to be able to edit text content in Froala.

In my mind it would make more sense for the "content" tab to be called "visual editor," then have another tab that's called "HTML," then another for the "Text" version of the email. The code tab could have an editor or a simple textarea box for raw code to be pasted in. Then either upon saving, or by detecting with Javascript, disable and ignore the Froala editor if the code box is not empty. Does that make sense?

**Comment by nickian** [comment]
Froala even strips out `<style>` tags and anything inside of them. This is incompatible with responsive emails because you need to have media queries to adjust your layout. Even the SendWithUs templates that are built in are broken because these styles have been removed.

**Comment by abooster** [comment]
I agree that if this isn't fixed, it would be a total dealbreaker. 
The WYSIWYG editor CANNOT be allowed to mess with the code that has been pasted as raw HTML code. 
And there's absolutely no reason for such a compromise. 
It is obviously possible to have a WYSIWYG editor and also be able to switch to raw HTML. Works in WordPress without any problems and without getting any code stripped. 

So, stripping or auto-changing of custom HTML = total dealbreaker! 

Oh, and let's not pretend that Mautic is meant for an average Mom and Pop shop owner. It's not! An average Mom and Pop shop owner (i.e. someone without heavy tech knowledge) won't be able to operate Mautic. But regardless of that, WordPress has both options WYSIWYG and raw HTML and it works without any problems there. So, has to work here as well.

**Comment by escopecz** [comment]
Ok, guys, thanks for your input. Now, how do we do that technically?

_And @abooster, please, try to keep the discussion calm. When you use phrases like "CANNOT" or "total dealbreaker!" or "let's not pretend", I feel like you're screaming at me which kind of discourages me to continue dealing with you and this issue._

**Comment by abooster** [comment]
@escopecz 
Screaming? Not screaming at all but rather **emphasising**. 
And reading through this thread I definitely felt that this emphasis was needed because you didn't appear to take this issue too seriously but rather trying to make a case for why it's not a big deal or that marketers would rather want to forego the HTML option. 

I am a marketer and I say it would be a dealbreaker if left unfixed. 
(can't imagine any other marketers wanting to allow their emails to be messed up)

**Comment by abooster** [comment]
It is definitely true that not all marketers want to learn HTML. 
But that's why tools like http://foundation.zurb.com/emails.html exist and are being used by marketers 
and that's why can't accept getting the HTML code messed up.

**Comment by escopecz** [comment]
OK, thanks for making your expressions clear. It's something a text comment is not good at. Now let's focus to the solution.

I created this issue (among others) when implementing Froala editor to Mautic:
https://github.com/froala/wysiwyg-editor/issues/1372

As you can see, they don't give a :hankey:  so I had to clear the table styles which Froala adds every time on before save. But I cannot fix code which Froala deletes if that's understandable.

Maybe if more people would bother them with issues about modifying the original source code, they'd listen.

I understand that this is a big issue. If you look at the rest of the GitHub issues, we have about 400 dealbreakers and limited resources to implement them all at short time. So I wasn't kidding when I asked for an easy technical solution for this issue. Because I can't think of any except the one I mentioned earlier.

**Comment by abooster** [comment]
OK, I think there is a simple solution, well, rather an acceptable workaround for this until Froala gets this issue fixed (or a better editor is found to replace Froala). 

So, I propose this as a temporary fix: 
Give the Mautic user an option to switch to raw HTML (i.e. without being able to use WYSIWYG). 
This is easy to implement and much better than a faulty Froala lockdown. 
I would personally love to have a  WYSIWYG option and I think that long-term it's definitely needed, but if having WYSIWYG means that my pasted-in HTML code gets messed up, then I'd rather forego  WYSIWYG completely and opt for a raw HTML box.

Ideally, the raw HTML box would have a preview link that opens a new window showing the preview.

**Comment by abooster** [comment]
@escopecz
Here, take a look at this: 
http://imgur.com/cDTe5La
I think that's a pretty good/acceptable solution.

**Comment by escopecz** [comment]
Well, as I wrote before, the HTML gets "cleared" by Froala every time the content is loaded to it. So switching it off and on won't prevent it.

**Comment by abooster** [comment]
I understand. The software in the screenshot above tells the user (underneath the box): 
"The WYSIWYG editor may try to correct your HTML code".

A more straightforward message could say something like:
"The WYSIWYG editor might change HTML code. So, if you decide to paste in raw HTML, don't toggle to the WYSIWYG editor."

That's good enough as a warning or as an explanation that to avoid unwanted changes in HTML the user shouldn't toggle to the WYSIWYG editor.
You just need to make sure that if the user toggles into raw HTML mode, Mautic remembers this setting and from now on makes it default when the user visits the edit page.

**Comment by escopecz** [comment]
Guys, check https://github.com/mautic/mautic/pull/1979 and let me know if there is something else you're missing. Some attribute which you use and gets removed?

**Comment by nickian** [comment]
At the bare minimum, these should be implemented in the app.js file where Froala is initialized on the textarea: `htmlAllowedTags: ['.*']` and  `htmlRemoveTags: ['']`

There are still a couple issues, but this let's us keep the `<style>` tags and the CSS within them.

**Comment by abooster** [comment]
As I am a marketer and not a developer (although I learn as I go and I can HTML) 
instructions like: 
"1) Apply this PR
2) Rebuild prod assets with app/console mautic:assets:generate command"
Those instructions are a little bit above my current ability to execute i.e. I lack the understanding to do that. 
I understand that Github is normally meant for developers who can do all those things. 
But I created an account here because I thought I need to report the bugs that I'm experiencing. (in fact, the error page in Mautic was linking to Github and asking to report the bug)

**Comment by nickian** [comment]
Not sure why you wouldn't just allow all tags.

@abooster, if you want to use my temporary solution, you can replace your app.js file with the one I have here: https://demo.contacts.live/media/js/app.js

This is the part of the code I modified:

```
{
    enter: mQuery.FroalaEditor.ENTER_BR,
    imageUploadURL: mauticBaseUrl + 's/file/upload',
    imageManagerLoadURL: mauticBaseUrl + 's/file/list',
    imageManagerDeleteURL: mauticBaseUrl + 's/file/delete',
    useClasses: false,
    htmlAllowedTags: ['.*'],
    htmlRemoveTags: [''],
    fullPage: true,
    lineBreakerTags: [''],
    linkAlwaysBlank: true
};
```

**Comment by escopecz** [comment]
@abooster there is a short tutorial if you'd like to learn how to test a PR: https://github.com/mautic/mautic#how-to-test-a-pull-request

But if that's too technical for you, don't worry about it.

@nickian I updated the PR with inspiration from your code. The only problem with it is that it allows all tags in all Froala editors in whole Mautic which is a bit dangerous. Style and Script tags from the editable content will be applied to the Mautic administration and it can break stuff. It should be applied only to the full-page editors which loads its content in an iframe and therefor isolated from Mautic. But I understand you mean it only as a quick fix of 2.0.0.

**Comment by abooster** [comment]
I just realized that I can just click the "Files changed" tab there and apply the appropriate changes to the affected file on my server. 
Well, this is something I can very easily do (as opposed to applying PR via the "official" route)
I'll try it as soon as possible and get back with my feedback.

**Comment by ldrrp** [comment]
Have these settings been set, https://www.froala.com/blog/styling-emails-with-wysiwyg-html-editors

**Comment by ldrrp** [comment]
I have this email that has an image set on a custom template that is now making it scale to 90% and its not even in the html preview, so i cant remove it.


https://www3.attotech.com/email/preview/110

**Comment by ldrrp** [comment]
I ended up recreating my entire template in froala. It sucked, but now at least it works better.

**Comment by nickian** [comment]
@escopecz thanks, that makes sense.

The stylesheet issue is resolved, but there is still a big issue with Froala add arbitrary line break tags in the content.

If possible, I think the best solution for the next release would be a separate HTML box for people who create their own custom designs. It could have some simple logic that says if there is content in the HTML box, ignore what's in Froala when hitting save. I would try to create this myself, but I'm not familiar enough with Mautic's code or Symfony.

**Comment by nickian** [comment]
Can you point me in the right direction? Where in the bundles can I find the email editor?

**Comment by escopecz** [comment]
@nickian sure. Any help is very welcome. Mautic has 2 modes. Production and development. The development mode shows PHP errors on the screen and also use source JS files instead of one minified JS file. To use the dev mode, pull Mautic from GH, install it and change the URL to use `index_dev.php` like this: `http://yourmautic.com/index_dev.php/s/`. The Froala editors are initiated in https://github.com/mautic/mautic/blob/staging/app/bundles/CoreBundle/Assets/js/1.core.js. The template for Email form is here: https://github.com/mautic/mautic/blob/staging/app/bundles/EmailBundle/Views/Email/form.html.php

Then there is a PHP class for the form itself, controller which handles the request, model which saves it and so on. You can find all of those classes in the EmailBundle.

**Comment by nickian** [comment]
@escopecz I'm a little lost. When try to go to http://localhost/mautic/index_dev.php/s, I get a 404 error.

**Comment by escopecz** [comment]
Do you use the GitHub version? I'm not sure that the dev environment is available in the production version.

**Comment by nickian** [comment]
Ah, that must be why. I cloned it from GitHub and it does work now. Thanks!

**Comment by palominoz** [comment]
Hi ! I applied as a patch this PR on my mautic installation, I did regen assets but still editor inserts arbitrary content and makes everything unusable. This has been something terrible to release! really! When is 2.1 coming out?  Will you patch with 2.0.2?

**Comment by nickian** [comment]
Yeah, Froala is just not going to work for me. It adds arbitrary `<br>` tags and has other issues. I'm having to paste my HTML directly into the database table row manually. Even after doing this, if I try to edit the subject line or from address, by initiating the editor, my code gets ruined. The ideal solution would be to have a separate, straight HTML box aside from this editor.

I'm trying to wrap my head around the code to see if I can create something like this. @escopecz do you know where I can find these lines that show up the `<head>` at in the source code?
`<script src="/app/bundles/CoreBundle/Assets/js/libraries/froala/froala_editor.js?v2.0.1" data-source="mautic"></script>`

**Comment by escopecz** [comment]
@nickian https://github.com/mautic/mautic/blob/staging/app/bundles/CoreBundle/Templating/Helper/AssetsHelper.php#L483

@palominoz make sure the code changes are really there in your Mautic and then you run the `php app/console mautic:assets:generate` command. Any help testing PRs before upcoming releases to make sure these issues won't be released is very welcome.

**Comment by palominoz** [comment]
@escopecz I applied changes and run command, but I am experiencing the same as @nickian at the moment.  Maybe it depends from the fact i dont have experience with mautic codebase but at the moment people needs a solution..

**Comment by nickian** [comment]
I think the PR only fixed the allowing of style and script tags. It didn't fix the issue where Froala still adds arbitrary code to the editor.

**Comment by escopecz** [comment]
Yes, it only enabled the style and script tags and it was already merged into 2.0.1 which is the latest version. I thought we are talking about https://github.com/mautic/mautic/pull/2032 now. If you have any questions about Froala editor, please, [ask them](https://github.com/froala/wysiwyg-editor/issues) first. If they say there is a solution, I can implement it.

**Comment by versadial** [comment]
I believe the editor also strips out width="200px" in an IMAGE. Styles appear to stay there, but when saving, if you have width defined outside of the style tag..it gets removed..and in outlook images appear full size.

**Comment by chandon** [comment]
To add width and height in image html tag, according to froala doc
https://www.froala.com/wysiwyg-editor/docs/options#imageOutputSize

1) the froala editor should be initialized like this (in file email.js)

    mQuery('#emailform_customHtml').froalaEditor({ imageOutputSize: true });

At this time, it's initialized like this

    mQuery('#emailform_customHtml').froalaEditor('popups.hideAll');

I'ts not compliant with the froala doc, what's popus.hideAll String for ?

2) according to https://www.froala.com/wysiwyg-editor/docs/methods#html.get you can get the HTML code inside froala with 

    mQuery('#emailform_customHtml').froalaEditor('html.get', true);

see https://github.com/mautic/mautic/blob/08d33e133cd7f5aff410d98073e9df0238ebef9b/app/bundles/CoreBundle/Assets/js/libraries/froala/plugins/image.js#L1713
width and height are only added with event html.beforeGet wich is triggered with html.get function only

**Comment by escopecz** [comment]
@chandon thanks for the comment! I added the imageOutputSize option here:

https://github.com/mautic/mautic/pull/2032/commits/e69e1ddcca6ca5bc17267200d1928c344ec61da1

To answer your questions:

1) That does not initialize a Froala editor. It tries to handle a bug in Froala reported here: https://github.com/froala/wysiwyg-editor/issues/1377

Froala for the full page editor is initialized here:
https://github.com/mautic/mautic/blob/staging/app/bundles/CoreBundle/Assets/js/1.core.js#L600

2) It doesn't have to be explicitly called, I tested that.

**Comment by escopecz** [comment]
@chandon Oh, I just noticed you submitted a PR with the imageOutputSize option... Now we have it in 2 PRs. Would you mind if I close yours in favor of https://github.com/mautic/mautic/pull/2032 which contains more editor fixes?

**Comment by chandon** [comment]
@escopecz Of course

**Comment by chandon** [comment]
@escopecz Could you please allow meta tag, http-equiv and content parameters in Froala (to specify utf8 for some webmail clients)

    htmlAllowedTags: ['meta', ...
    htmlAllowedAttrs: ['http-equiv', 'content', ...

**Comment by escopecz** [comment]
@chandon it should be done by https://github.com/mautic/mautic/pull/2032/files#diff-c39ac84ffcabd2815b713f76d024a3c1R575

**Comment by chandon** [comment]
@escopecz it should, but it's not working... if you want options.htmlAllowedTags override Mautic.basicFroalaOptions, you have to use 

    mQuery.extend(Mautic.basicFroalaOptions, options)

and not 

    mQuery.extend(options, Mautic.basicFroalaOptions)

Please see https://api.jquery.com/jquery.extend/

**Comment by escopecz** [comment]
@chandon good catch! Fixed in https://github.com/mautic/mautic/pull/2032/commits/70c8f37612909e4a0eb9e050054eda5f5c65b472

**Comment by 750430** [comment]
Hi ... any chance this could be deployed to a Mautic-hosted instance?

**Comment by mingo3369** [comment]
I ran into this issue and appreciate that it is already being addressed by you folks.  The work around I am using is not to preview the text in the editor.  I just paste it and save and close the editor.  This seems to work.  Maybe you guys already knew that.

Again thanks for your efforts on this issue.

**Comment by escopecz** [comment]
I'll close this now, because #2258 was merged and release in Mautic 2.1.1. Please, if you still get some stripping issues, create a new issue with the code which gets stripped.

**Comment by NaveenPSingh** [comment]
In the same way I am creating a custom template in mautic and if I just select template and apply then page view is perfect but If I select template, open builder then apply, page view breaks like:
 ` if ($(document).scrollTop() > 50) { `
converted in builder mode
`  if ($(document).scrollTop() &gt; 50) {`
then after apply page view converted in
`  if ($(document).scrollTop() &amp;gt; 50) {`
I used all script between **{% raw %}** ----- **{% endraw %}**
